src/resources/rmt_esp32.cc

```diff
@@ -0,0 +1,218 @@
+    RMT_CHANNEL_4, RMT_CHANNEL_5, RMT_CHANNEL_6, RMT_CHANNEL_7
+);
+
+class RMTResourceGroup : public ResourceGroup {
+ public:
+  TAG(RMTResourceGroup);
+  RMTResourceGroup(Process* process)
+    : ResourceGroup(process, null) { }
+
+  virtual void on_unregister_resource(Resource* r) {
+    rmt_channel_t channel = static_cast<rmt_channel_t>(static_cast<IntResource*>(r)->id());
+    rmt_driver_uninstall(channel);
+    rmt_channels.put(channel);
+  }
+
+};
+
+MODULE_IMPLEMENTATION(rmt, MODULE_RMT);
+
+PRIMITIVE(init) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  RMTResourceGroup* rmt = _new RMTResourceGroup(process);
+  if (!rmt) MALLOC_FAILED;
+
+  proxy->set_external_address(rmt);
+  return proxy;
+}
+
+PRIMITIVE(use) {
+  ARGS(RMTResourceGroup, resource_group, int, channel_num)
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+
+  if (!rmt_channels.take(channel_num)) ALREADY_IN_USE;
+
+  // TODO install RMT driver for channel?
+  IntResource* resource = resource_group->register_id(channel_num);
+  if (!resource) {
+    rmt_channels.put(channel_num);
+    MALLOC_FAILED;
+  }
+  proxy->set_external_address(resource);
+
+  return proxy;
+}
+
+PRIMITIVE(unuse) {
+  ARGS(RMTResourceGroup, resource_group, IntResource, resource)
+  int channel = resource->id();
+  resource_group->unregister_id(channel);
+  resource_proxy->clear_external_address();
+  return process->program()->null_object();
+}
+
+esp_err_t configure(const rmt_config_t* config, rmt_channel_t channel_num, size_t rx_buffer_size, Process* process) {
+  esp_err_t err = rmt_config(config);
+  if (ESP_OK != err) return err;
+
+  err = rmt_set_source_clk(channel_num, RMT_BASECLK_APB);
+  if (ESP_OK != err) return err;
+
+  err = rmt_driver_install((rmt_channel_t) channel_num, rx_buffer_size, 0);
+  if (ESP_OK != err) return err;
+  return err;
+}
+
+PRIMITIVE(config_tx) {
+  ARGS(int, pin_num, int, channel_num, int, mem_block_num, int, clk_div, int, flags,
+       bool, carrier_en, int, carrier_freq_hz, int, carrier_level, int, carrier_duty_percent,
+       bool, loop_en, bool, idle_output_en, int, idle_level)
+
+  rmt_config_t config = { };
+
+  config.gpio_num = (gpio_num_t) pin_num;
+  config.channel = (rmt_channel_t) channel_num;
+  config.mem_block_num = mem_block_num;
+  config.clk_div = clk_div;
+  config.flags = flags;
+  config.rmt_mode = RMT_MODE_TX;
+  rmt_tx_config_t tx_config = { 0 };
+  tx_config.carrier_en = carrier_en;
+  tx_config.carrier_freq_hz = carrier_freq_hz;
+  tx_config.carrier_level = (rmt_carrier_level_t) carrier_level;
+  tx_config.carrier_duty_percent = carrier_duty_percent;
+  tx_config.loop_en = loop_en;
+  tx_config.idle_output_en = idle_output_en;
+  tx_config.idle_level = (rmt_idle_level_t) idle_level;
+  config.tx_config = tx_config;
+
+  esp_err_t err = configure(&config, (rmt_channel_t) channel_num, 0, process);
```

Comment:
The documentation suggests that you can update configuration values one by one.
However, by not setting the rx_config it's not clear to me whether it would be overwritten.

Ideally, setting the rx should not destroy the tx.

Reply:
It would be helpful with a reference to the documentation in question.

Reply:
The [documentation](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#configure-driver) says to configure RMT and then install the driver. It is unclear to me whether dynamic updates to the configuration are supported.

A quick inspection of the code suggests that rx configuration does not overwrite tx specific parameters, but general parameters are overwritten.